Large Langage Model for Genomics using Python (#5962)
galaxyproject/training-material: topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md

from sklearn.decomposition import PCA
from transformers import pipeline
from typing import List, Tuple
```

> <comment-title>Versions</comment-title>
>
> This tutorial has been tested with following versions:
> - `numpy` > 1.26.4
> - `transformers` > 4.47.1
> 
> You can check the versions with:
>
> ```python
> np.__version__
> transformers.__version__
> ```
{: .comment}

## Check and configure available resources

Let's check the GPU usage and RAM:

```python
!nvidia-smi
```

Let's configure PyTorch and the CUDA environment -- software and hardware ecosystem provided by NVIDIA to enable parallel computing on GPU -- to optimize GPU memory usage and performance:

1. Enables CuDNN benchmarking in PyTorch:

    ```python
    torch.backends.cudnn.benchmark=True
    ```

2. Set an environment variable that configures how PyTorch manages CUDA memory allocations

    ```python
    os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "max_split_size_mb:32"
    ```

# Generate Synthetic DNA Sequences

Using the pre-trained language model which was trained on 1,011 *Saccharomyces cerevisiae* isolates, we would like to generate 100 synthetic yeast DNA sequences.

## Build the Sequence Generator

> First, we need set up the sequence generation pipeline. This pipeline will enable us to generate synthetic DNA sequences that mimic natural genomic sequences. By leveraging the power of language models, we can create novel DNA sequences for various applications in synthetic biology.

Reviewer: ```suggestion
First, we need to set up the sequence generation pipeline. This pipeline will enable us to generate synthetic DNA sequences that mimic natural genomic sequences. By leveraging the power of language models, we can create novel DNA sequences for various applications in synthetic biology.
```

        for kmer, count in kmer_dict.items():
            if kmer in kmer_index:
                matrix[seq_idx, kmer_index[kmer]] = count

    return matrix, all_kmers
```

Let's now count 4-mers for artificial and real sequences.

```python
k = 4
artificial_seq_kmer_counts, kmers = kmer_counts_matrix(artificial_sequences, k=k)
real_seq_kmer_counts, kmers = kmer_counts_matrix(real_sequences, k=k)
```

> <question-title></question-title>
>
> 1. What is the format of `artificial_seq_kmer_counts` and `real_seq_kmer_counts`?
> 2. What are the dimensions?
>
> > <solution-title></solution-title>
> >
> > 1. `artificial_seq_kmer_counts` and `real_seq_kmer_counts` are 2-matrices.
> > 2. 100 x 256 ($$256 = 4^{4}$$)
> >
> {: .solution}
>
{: .question}

To visualize and compare real and generated DNA sequences in a continuous space, we can utilize Principal Component Analysis (PCA). PCA is a powerful dimensionality reduction technique that transforms high-dimensional data into a lower-dimensional space while preserving as much variability as possible. This allows us to visualize the similarities and differences between the sequences effectively.

```python

> def plot_pca_projection(set1: np.ndarray, set2: np.ndarray, title: str = "PCA Projection") -> None:

Reviewer: Minor suggestion: Instead of PCA for showing projections, UMAP or tSNE could also be used.

> >     """
> >     # Initialize the sequence generation pipeline
> >     generator = pipeline("text-generation", model=model_name)
> >     # Generate synthetic DNA sequences
> >     synthetic_dna = generator(
> >        "",
> >        max_length=max_length,
> >        do_sample=True,
> >        top_k=top_k,
> >        temperature=temp,
> >        repetition_penalty=repetition_penalty,
> >        num_return_sequences=num_sequences,
> >        eos_token_id=0,
> >     )
> >     # Extract and clean the generated sequences
> >     artificial_sequences = [seq["generated_text"].replace(" ", "") for seq in synthetic_dna]
> >     return artificial_sequences
> >
> >
> > temperatures = [0.001, 0.01, 0.1, 0.5, 1.0, 1.5]
> > for temp in temperatures:
> >     art_sequences = generate_synthetic_dna(model_name, max_length,  num_sequences, temp)
> >     art_seq_kmer_counts, kmers = kmer_counts_matrix(art_sequences, k=k)
> >     plot_pca_projection(real_seq_kmer_counts, art_seq_kmer_counts)
> >     var = np.mean(np.var(art_seq_kmer_counts, axis=0))
> >     print(f"Variance for temperature of { temp }: { var }")
> > ```
> {: .solution}
>
{: .hands-on}

> <question-title></question-title>
>

> > 1. How similar look like the 5 first generated DNA sequences given the temparature?

Reviewer 2: ```suggestion
> 1. How similar do the 5 first generated DNA sequences look like given the temperature?
```

Let's start with a synthetic DNA sequence generated with a low temperature setting. Low temperature settings reduce variability, making the generated sequences more similar to real sequences. This step helps ensure that the sequence is biologically plausible and likely to have counterparts in existing databases.

> <hands-on-title></hands-on-title> 
>
> 1. Generate the synthetic DNA sequences with a temperature value of $$0.01$$
> 2. Get the first sequence
> 3. Make a [Standard Nucleotide BLAST search](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome) of the first sequence
>
> > <solution-title></solution-title>
> > 
> > ```python
> > 
> > art_sequences = generate_synthetic_dna(model_name, max_length,  num_sequences, 0.01)
> > art_sequences[0]
> > ```
> {: .solution}
>
{: .hands-on}

> <question-title></question-title>
>

> > 1. How many significant similarity have been found?

Reviewer 2: ```suggestion
> 1. How many significant similarities have been found?
```

> >
> {: .solution}
>
{: .question}

When generating sequences with low temperature, we observed limited variability among the sequences. This lack of variability can be useful for ensuring consistency but may limit the exploration of novel sequence spaces. To introduce more variability and potentially discover even more novel sequences, we can increase the temperature setting during sequence generation. Higher temperatures introduce greater randomness, leading to more diverse and potentially unique sequences.

> <hands-on-title></hands-on-title> 
>
> 1. Generate the synthetic DNA sequences with a temperature value of $$1$$
> 2. Get the first sequence
> 3. Make a [Standard Nucleotide BLAST search](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome) of the first sequence
>
> > <solution-title></solution-title>
> > 
> > ```python
> > 
> > art_sequences = generate_synthetic_dna(model_name, max_length,  num_sequences, 1)
> > art_sequences[0]
> > ```
> {: .solution}
>
{: .hands-on}

> <question-title></question-title>
>

> > 1. How many significant similarity have been found?

Reviewer 2: ```suggestion
> 1. How many significant similarities have been found?
```

```python
def extract_and_translate_orf(
    sequence: str,
    orfs: List[Tuple[int, int, str, str]],
    orf_index: int = 0,
) -> str:
    """
    Extract an ORF from a DNA sequence and translate it into a protein sequence.

    Parameters:
    - sequence (str): The DNA sequence containing the ORF.
    - orfs (List[Tuple[int, int, str, str]]): List of detected ORFs with start, stop positions, strand, and description.
    - orf_index (int): Index of the ORF to extract and translate. Default is 0.

    Returns:
    - str: The translated protein sequence.
    """
    # Extract the ORF sequence using the start and stop positions
    start, stop, strand, description = orfs[orf_index]
    orf_sequence = sequence[start:stop]

    # Translate the ORF sequence into a protein sequence
    protein_sequence = str(Seq(orf_sequence).translate())

    return protein_sequence
```

For the first detected ORFs:

```python
protein_seq = extract_and_translate_orf(seq, orfs)
print(f"Translated Protein Sequence: {protein_seq}")
```

> After translating a synthetic DNA sequence into a protein sequence, the next crucial step is to analyze this sequence to uncover its potential biological function. This involves identifying known protein domains or motifs within the sequence that may indicate specific functions, such as binding sites or active regions. Verifying the biological plausibility of the translated sequence is essential, which can be achieved by comparing it to known protein sequences in databases like UniProt or GenBank using tools like BLAST. Additionally, examining the conservation of the sequence across different species can provide insights into its functional importance. If the sequence shows potential for biological function, it can serve as a starting point for further research, including structural modeling to predict its three-dimensional structure, functional assays to test its activity, and evolutionary studies to understand its origins and adaptations. This comprehensive analysis not only enhances our understanding of the generated sequences but also opens new avenues for scientific exploration and application in synthetic biology and genetic engineering.

Author: ```suggestion
After translating a synthetic DNA sequence into a protein sequence, the next crucial step is to analyze this sequence to uncover its potential biological function. This involves identifying known protein domains or motifs within the sequence that may indicate specific functions, such as binding sites or active regions. Verifying the biological plausibility of the translated sequence is essential, which can be achieved by comparing it to known protein sequences in databases like UniProt or GenBank using tools like InterProScan. Additionally, examining the conservation of the sequence across different species can provide insights into its functional importance. If the sequence shows potential for biological function, it can serve as a starting point for further research, including structural modeling to predict its three-dimensional structure, functional assays to test its activity, and evolutionary studies to understand its origins and adaptations. This comprehensive analysis not only enhances our understanding of the generated sequences but also opens new avenues for scientific exploration and application in synthetic biology and genetic engineering.
```

Author: Better?

Reviewer 2: You could leave BLAST as well: "...using tools like BLAST and InterproScan."

Author: Updated

Reviewer 3: Can we please avoid useing BLAST for protein sequences, we should really use DIAMOND here. Its so much faster and does not mess with the cluster.

> After translating a synthetic DNA sequence into a protein sequence, the next crucial step is to analyze this sequence to uncover its potential biological function. This involves identifying known protein domains or motifs within the sequence that may indicate specific functions, such as binding sites or active regions. Verifying the biological plausibility of the translated sequence is essential, which can be achieved by comparing it to known protein sequences in databases like UniProt or GenBank using tools like BLAST. Additionally, examining the conservation of the sequence across different species can provide insights into its functional importance. If the sequence shows potential for biological function, it can serve as a starting point for further research, including structural modeling to predict its three-dimensional structure, functional assays to test its activity, and evolutionary studies to understand its origins and adaptations. This comprehensive analysis not only enhances our understanding of the generated sequences but also opens new avenues for scientific exploration and application in synthetic biology and genetic engineering.

Author: ```suggestion
After translating a synthetic DNA sequence into a protein sequence, the next crucial step is to analyze this sequence to uncover its potential biological function. This involves identifying known protein domains or motifs within the sequence that may indicate specific functions, such as binding sites or active regions. Verifying the biological plausibility of the translated sequence is essential, which can be achieved by comparing it to known protein sequences in databases like UniProt or GenBank using tools like Diamond and InterProScan. Additionally, examining the conservation of the sequence across different species can provide insights into its functional importance. If the sequence shows potential for biological function, it can serve as a starting point for further research, including structural modeling to predict its three-dimensional structure, functional assays to test its activity, and evolutionary studies to understand its origins and adaptations. This comprehensive analysis not only enhances our understanding of the generated sequences but also opens new avenues for scientific exploration and application in synthetic biology and genetic engineering.
```
